### ISSUE TITLE
gpu memory leakage into another one

### ISSUE THREAD

**Issue by yanaiela** [allenai/allennlp#1002]
hey,

I'm trying to use the full gpus capacity, so I'm running several processes in parallel (several gpus on the same machine).
After running the second process, I noticed it uses some memory from the first allocated gpu (and causing it to run out of memory after a while).

This is how I run the process:

> python -m allennlp.run predict \
https://s3-us-west-2.amazonaws.com/allennlp/models/elmo-constituency-parser-2018.03.14.tar.gz \
--cuda-device 1 --batch-size 64 sentences_01 --predictor=constituency-parser > sentences_01.pred

(with cuda-device 0 for the first one)

is this the correct way of running it?

**Comment by DeNeutoy** [comment]
Hey, sorry about that - we aren't as strict as we should be about multi device contexts. If you know why this is happening, (normally it will be somewhere where we call `.cuda()` without specifying a device) please submit a PR. In the short term though, there is an easy fix - you can set the `CUDA_VISIBLE_DEVICES` environment variable separately in each process. This will prevent the leakage.

**Comment by yanaiela** [comment]
Thanks!

**Comment by yanaiela** [comment]
Oh and there's another thing, it might be another issue though.
I'm running the model on a relatively large file (~180M with 3.5 Milion sentences).
The memory that the process use keeps getting bigger and bigger and it ends by crushing due to lack of memory.

**Comment by DeNeutoy** [comment]
CPU or GPU? I need more information to help, e.g how fast does it grow, is it only if you use GPU prediction etc.

**Comment by yanaiela** [comment]
only gpu (like the command I posted here before)
regarding the speed, I think it took around 4-5 hours to reach the 11G memory limit.

what do you mean by gpu prediction?

+ now that I ran several in parallel, I see that they don't grow in equal pace. I've started 4 processes at the same time (~1 hour ago) and this is the current memory usage:
4595M, 9051M, 4103M, 10409M

**Comment by DeNeutoy** [comment]
Are you seeing sudden jumps in memory usage? What's likely happening is that you are incrementally encountering longer sentences in your large dataset, which increase the memory allocated for a batch. Do you have particularly long sentences in your dataset? 

A way you can test this is to sort your dataset by length and process it longest first (or use the `--overrides` flag to pass the `biggest_batch_first` argument to the [iterator class](https://github.com/allenai/allennlp/blob/master/allennlp/data/iterators/bucket_iterator.py#L43)), which will make the predictor crash immediately if this is what is happening.

**Comment by yanaiela** [comment]
as you said, it seems to be the problem.
thanks!
